src/test/app/TrustAndBalance_test.cpp

```diff
@@ -0,0 +1,512 @@
+                }));
+
+            BEAST_EXPECT(wsc->invoke("unsubscribe",jv)[jss::status] == "success");
+        }
+    }
+
+    void
+    testWithPath ()
+    {
+        testcase ("Payments With Xfer Fee");
+        using namespace test::jtx;
+
+        Env env {*this};
+        Account gw {"gateway"};
+        Account alice {"alice"};
+        Account bob {"bob"};
+
+        env.fund (XRP(10000), gw, alice, bob);
+        env.close();
+
+        // set a transfer rate
+        env (rate (gw, 1.1));
+
+        env (trust (alice, gw["AUD"](100)));
+        env (trust (bob, gw["AUD"](100)));
+
+        env (pay (gw, alice, alice["AUD"](4.4)));
+        env.require (balance (alice, gw["AUD"](4.4)));
+
+        // alice sends gw issues to bob with a max spend that allows for the
+        // xfer rate
+        env (pay (alice, bob, gw["AUD"](1)), sendmax(gw["AUD"](1.1)));
+        env.require (balance (alice, gw["AUD"](3.3)));
+        env.require (balance (bob, gw["AUD"](1)));
+
+        // alice sends bob issues to bob with a max spend
+        env (pay (alice, bob, bob["AUD"](1)), sendmax(gw["AUD"](1.1)));
+        env.require (balance (alice, gw["AUD"](2.2)));
+        env.require (balance (bob, gw["AUD"](2)));
+
+        // alice sends gw issues to bob with a max spend
+        env (pay (alice, bob, gw["AUD"](1)), sendmax(alice["AUD"](1.1)));
+        env.require (balance (alice, gw["AUD"](1.1)));
+        env.require (balance (bob, gw["AUD"](3)));
+
+        // alice sends bob issues to bob with a max spend in alice issues.
+        // expect fail since gw is not involved
+        env (pay (alice, bob, bob["AUD"](1)), sendmax(alice["AUD"](1.1)),
+            ter(tecPATH_DRY));
+
+        env.require (balance (alice, gw["AUD"](1.1)));
+        env.require (balance (bob, gw["AUD"](3)));
+    }
+
+    void
+    testIndirect ()
+    {
+        testcase ("Indirect Payment");
+        using namespace test::jtx;
+
+        Env env {*this};
+        Account gw {"gateway"};
+        Account alice {"alice"};
+        Account bob {"bob"};
+
+        env.fund (XRP(10000), gw, alice, bob);
+        env.close();
+
+        env (trust (alice, gw["USD"](600)));
+        env (trust (bob, gw["USD"](700)));
+
+        env (pay (gw, alice, alice["USD"](70)));
+        env (pay (gw, bob, bob["USD"](50)));
+
+        env.require (balance (alice, gw["USD"](70)));
+        env.require (balance (bob, gw["USD"](50)));
+
+        // alice sends more than has to issuer: 100 out of 70
+        env (pay (alice, gw, gw["USD"](100)), ter(tecPATH_PARTIAL));
+
+        // alice sends more than has to bob: 100 out of 70
+        env (pay (alice, bob, gw["USD"](100)), ter(tecPATH_PARTIAL));
+
+        env.close();
+
+        env.require (balance (alice, gw["USD"](70)));
+        env.require (balance (bob, gw["USD"](50)));
+
+        // send with an account path
+        env (pay (alice, bob, gw["USD"](5)), test::jtx::path(gw));
```

Comment:
I believe if the no ripple (tfNoRippleDirect) flag is set, the transaction would fail without a specified path.

`env(pay(alice, bob, gw["USD"](5)), txflags(tfNoRippleDirect), test::jtx::path(gw))`